src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -0,0 +1,185 @@
+
+    # ... wait for 5 minutes until the buffer is filled
+    await asyncio.sleep(5)
+
+    # return an numpy array from the window
+    a = window[time_start:time_end]
+    # and use it to for example calculate the mean
+    mean = a.mean()
+    '''
+
+    **Example2** (create a polars data frame from a `MovingWindow`):
+
+    ```
+    import polars as pl
+
+    # create a window that stores two days of data
+    # starting at 1.1.23 with samplerate=1
+    window = MovingWindow(size = (60 * 60 * 24 * 2), sample_receiver)
+
+    # wait for one full day until the buffer is filled
+    asyncio.sleep(60*60*24)
+
+    # create a polars series with one full day of data
+    time_start = datetime(2023, 1, 1)
+    time_end = datetime(2023, 1, 2)
+    s = pl.Series("Jan_1", mv[time_start:time_end])
+    ```
+    """
+
+    def __init__(
+        self,
+        size: int,
+        resampled_data_recv: Receiver[Sample],
+        sampling_period: timedelta,
+        window_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """
+        Initialize the MovingWindow.
+
+        This method creates the underlying ringbuffer and starts a
+        new task that updates the ringbuffer with new incoming samples.
+        The task stops running only if the channel receiver is closed.
+
+        Args:
+            size: The number of elements that are stored.
+            resampled_data_recv: A receiver that delivers samples with a
+                given sampling period.
+            sampling_period: The sampling period.
+            window_alignment: A datetime object that defines a point in time to which
+                the window is aligned to modulo window size.
+                (default is midnight 01.01.01)
+                For further information, consult the class level documentation.
+
+        Raises:
+            asyncio.CancelledError: when the task gets cancelled.
+        """
+        self._resampled_data_recv = resampled_data_recv
+        self._buffer = OrderedRingBuffer(
+            np.empty(shape=size, dtype=float),
+            sampling_period=sampling_period,
+            time_index_alignment=window_alignment,
+        )
+        self._copy_buffer = False
+        self.update_window_task: asyncio.Task = asyncio.create_task(self._run_impl())
+        log.debug("Cancelling MovingWindow task: %s", __name__)
+
+    async def _run_impl(self) -> None:
+        """Awaits samples from the receiver and updates the underlying ringbuffer."""
+        try:
+            async for sample in self._resampled_data_recv:
+                log.debug("Received new sample: %s", sample)
+                self._buffer.update(sample)
+        except asyncio.CancelledError:
+            log.info("MovingWindow task has been cancelled.")
+
+        log.error("Channel has been closed")
+
+    def stop(self) -> None:
+        """Cancel the running task and stops the MovingWindow."""
+        self.update_window_task.cancel()
```

Comment:
Sorry I didn't see the task is actually a public member, in that case I don't think it is necessary to expose a `stop()`, sorry! If you keep it it should probably be `async` and await on `join()` too to make it really useful, but I'm fine (and maybe it is best) to just expose the task.

This is actually something that we should agree on because it comes up each time we create tasks, right now we have no unified approach to this and it is starting to be a mess IMHO.

Reply:
It wasn't public and you are right that's something we should agree on. I'm in favor of having a method with a common name like stop.

Reply:
I'll make the task private again and leave the stop method.

Reply:
:+1: you might want to use the convenience `frequenz.sdk._internal.asyncio.cancel_and_await()` function.

Reply:
> If you keep it it should probably be async and await on join() too to make it really useful

Can you please elaborate? What should be awaited and why does awaiting it make it more useful?

Reply:
Sure. So I might mention things that you already know or are obvious, please bare with me :)

`cancel()` is sync, so it just tells the task to stop the next time it can (usually when it is `await`ing something).

If you need to wait for the task to finish (for example, before exiting a `test_xxx` function you want all the tasks to finish to avoid errors about tasks still running), you need to call `join()` (this is also true for normal termination, not just cancellation).

If the `update_window_task` member is made private again, then the user have no way to do the joining via a public interface.

So I see 2 approaches. Mimic the `Task` object and make the `MovingWindow` class provide a sync `cancel()` plus async `join()` or simplify it and use an async `stop()` that does both.

I'm not really convinced about what's best but I have the feeling that, even when it might be more work, going for a `Task`-compatible interface will be more future proof and more familiar to Python users. Maybe we could provide the extra `stop()` on top of it for convenience (maybe it can be a simple function like `cancel_and_await()`).

I wonder if we should add a `Task` object to the SDK from which any other objects that spawn tasks (like actors) should inherit. It could also provide much of the boilerplate too, like we could just have an internal list of tasks and the `Task.cancel()` will automatically call `cancel()` on all of them, and so on with the other methods.

OK, I diverged too much here, I'll probably create a new discussion from this comment :D

Reply:
> OK, I diverged too much here, I'll probably create a new discussion from this comment :D

yes please